### PR TITLE
no python in shell scripts

### PR DESCRIPTION
**Pull request by onur-ozkan**
Helps with #94829.

For now this needs you to manually opt in with `cargo run --manifest-path src/bootstrap/Cargo.toml -p bootstrap-shim`.

### CODE REVIEW

**Comment by ozkanonur** [comment]
Regarding to https://github.com/rust-lang/rust/pull/102282#issuecomment-1413129984

> Right now it doesn't actually save any time, because the shim depends on all of bootstrap's dependencies, but I think we can cut down the shim's dependencies to essentially 0 in a follow-up.

Since this is one of the long-term defined task, I believe this should be done in a seperated PR as a seperated task. `bootstrap-shim` is currently built on top of `MinConfig` and `Flags` which are defined in `bootstrap` tree.

> Right now bootstrap-shim doesn't accept any arguments (oops):

This should work just fine with this PR.

> This is non-trivial to fix with the current arg-parsing setup: either we'd have to share flags.rs between bootstrap and bootstrap-shim (which I'd rather not do to cut down on compile times), or we'd have to switch to a different arg-parsing library.

I used `Flags` in bootstrap shim, not just because of fixing the arg passing bug, but also for easier way to set `MinConfig` specific flags/options without parsing flags from arguments twice. We already parse flags in `Config::parse` and we need them in `MinConfig::parse` as well.


> We should add a download-bootstrap # bool option to config.toml to tell bootstrap-shim to fall back to using python instead.

I am not sure if this should be covered in this PR.

cc @jyn514

**Comment by ozkanonur** [comment]
ready @rustbot

**Comment by albertlarsan68** [comment]
The correct syntax is @rustbot ready, not the other way around.

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (albertlarsan68) [inline src/bootstrap/bin/bootstrap-shim.rs:6]: This is a bit of a misnomer, because what it really does is running the shell scripts, not python directly.

**Review by albertlarsan68 — COMMENTED** [review]
Not a full review

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (jyn514) [reply src/bootstrap/bin/bootstrap-shim.rs:6]: oh, that's a problem, they're going to call each other in an infinite loop. We need to add back the code that runs python first before we can merge this PR I think.

cc https://github.com/rust-lang/rust/issues/107907

File: src/tools/x/Cargo.toml
Comment (jyn514) [inline src/tools/x/Cargo.toml:3]: this doesn't actually need to be updated, the *behavior* of x isn't changing, only the code

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (jyn514) [inline src/bootstrap/bin/bootstrap-shim.rs:18]: This still has the problem where we end up infinitely calling the shell scripts recursively. Can you change `src/tools/x` to refactor the python search into a different function from the shell script search, so we can *only* look for python?

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (jyn514) [inline src/bootstrap/bin/bootstrap-shim.rs:27]:
```suggestion
        &["src/bootstrap", "src/tools/build_helper", "src/tools/x"],
```

File: src/bootstrap/dist.rs
Comment (jyn514) [inline src/bootstrap/dist.rs:2178]: this comment is outdated
```suggestion
```

File: src/bootstrap/config.rs
Comment (jyn514) [inline src/bootstrap/config.rs:385]: Please make this `pub(crate)` so it can only be used from `minimal_config`.

**Review by jyn514 — COMMENTED** [review]
This broadly looks ok :) I have several overall comments:

- I want to actually test this out; I'm going to do a try build shortly since right now basically the only thing that's running is `x::main` which runs the python script.
- I am *really* nervous that we don't have tests for `Config`. If there's a way to dump `Config` when run through the shim and when run directly through python, it would make me a *lot* more comfortable.
- I wrote most of this code originally and I'd like Mark to review it when he gets a chance; but I want to fix the other parts of this first so he's not reviewing code that's about to change.

File: src/bootstrap/config.rs
Comment (ozkanonur) [reply src/bootstrap/config.rs:385]: This was for the `src/bootstrap/config/tests.rs` refactoring. I will refactor it again and keep this struct private.

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (ozkanonur) [reply src/bootstrap/bin/bootstrap-shim.rs:18]: Will do it

**Comment by ozkanonur** [comment]
> I am really nervous that we don't have tests for Config. If there's a way to dump Config when run through the shim and when run directly through python, it would make me a lot more comfortable.

We definetly need testing for this. I will figure out a way to do this. Other than that your notes should be fixed with https://github.com/rust-lang/rust/pull/107812/commits/1050f7d5035382af8a2363107e6df75cb665685f

File: src/tools/x/src/main.rs
Comment (ozkanonur) [inline src/tools/x/src/main.rs:157]: Do we want this for `bootstrap-shim` ?

Comment (jyn514) [reply src/tools/x/src/main.rs:157]: No, it's only there so tidy can warn on an outdated version, which doesn't matter for the shim.

**Review by jyn514 — APPROVED** [review]
ok, some overall comments:

1. right now `MinimalConfig::unpack` is broken when `--dry-run` is set. I'm not sure we actually need to fix that here - this PR has already been in limbo a while and it seems ok to fix these minor bugs in follow-ups.
2. this works!! I hard-coded it to download the pre-compiled version in `9c780369554efa7924053b88d1e534deeed0092e` and it was able to build the standard library :)
3. I don't think we need to block this on testing that Config behaves exactly the same way, since `bootstrap-shim` is opt-in :)

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (jyn514) [inline src/bootstrap/bin/bootstrap-shim.rs:22]:
```suggestion
    println!("Running pre-compiled bootstrap binary");
    Command::new(bootstrap_bin).args(args).status().expect("failed to spawn bootstrap binairy");
```

**Comment by ozkanonur** [comment]
> 2. this works!! I hard-coded it to download the pre-compiled version in `9c780369554efa7924053b88d1e534deeed0092e` and it was able to build the standard library :)

Amazing :)

According to 1 and 3, I will handle them in individual PRs once this one gets merged

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/bootstrap-shim.rs:47]: This code looks near-identical to the x tool's find_and_run_available_bootstrap_script -- why do we limit ourselves to py here? Can we add a comment on this function and that function explaining the duplication?

File: src/bootstrap/min_config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/min_config.rs:20]: I think we should embed rationale here (i.e. in this comment or in a doc comment in this file) for:

* Why we have minimal vs. non-minimal config, if it all lives in bootstrap anyway.
* What the process for reviewers/authors should be when deciding where to put something.
  * What, if any, are implications of that choice? How permanent is it, in terms of breaking users if we move from one to the other?

(This will help me review the code as well).

File: src/bootstrap/download.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/download.rs:540]: It looks like a bunch of code here is maybe just moved, but GitHub is showing it as new code -- can we adjust the diffs or split into distinct commits or otherwise structure this so that this is not new code in diffs?

**Review by Mark-Simulacrum — COMMENTED** [review]
We'll also want to remove the merge commit from history just as an FYI.

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs:71]: There are helper functions created in there to avoid some of the duplicated logics

File: src/bootstrap/download.rs
Comment (ozkanonur) [reply src/bootstrap/download.rs]: Should be fine now

File: src/bootstrap/bin/bootstrap-shim.rs
Comment (ozkanonur) [reply src/bootstrap/bin/bootstrap-shim.rs]: > Where we end up infinitely calling the shell scripts recursively.

Can you clarify how this can happen? All I see it will always execute `x` script, which will look for python version and then execute `x.py`

File: src/bootstrap/min_config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/min_config.rs:21]: Can we avoid "BC" here? I assume that means "build caching", but it does not feel like a standard term to me.

File: src/bootstrap/min_config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/min_config.rs:22]: "if possible" feels like it's not very clear to me. For example, I see comments below about verbose and dry_run *not* being used in the shim, yet they are here. So that seems odd - why are they here?

File: src/bootstrap/config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:749]: This feels like the wrong approach to me. We should *always* start with MinimalConfig::parse, and then take that as an argument for the big Config struct. Any processing that is necessary for the minimal config is likely to *also* be necessary here, and can be read from fields in that struct as needed.

I think that will help remove the helper functions shared between the two (e.g., set_cfg_path_and_return_toml_cfg) and make the relationship between the two clearer in general.

File: src/bootstrap/config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:977]: This doesn't look like it needs to be reordered with the profile merging? Can we avoid spurious changes in this file?

File: src/bootstrap/config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:805]: This feels like a significant change that shouldn't be coupled to factoring out a minimal configuration - do we need it in this PR? (Can we split this PR up?)

How did this used to work? I'm not seeing the previous call to this...

**Review by Mark-Simulacrum — COMMENTED** [review]
I get the feeling that this PR is trying to do too much which both makes it harder to review and as such harder to land. But given the very sparse PR description, it's not clear to me what the minimal step this PR is taking is; it looks like it (roughly speaking):

* Introduces a notion of a minimal configuration
* Adds a new component to be distributed (bootstrap-shim)
* Refactors a bunch of code (mostly I think a result of minimal config, but seemingly not only - parts feel independent).

I'm having a hard time with this coupling, is there a chance we can avoid it? I've left a few comments inline that I think will help factor things out a little.

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs:21]: It means "backward compatibility". Pretty common and popular term. I will explicitly update it as "backward compatibility".

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs:22]: I believe it's clear enough. We simply say "the reason we have this minimal version is to avoid breaking backward compatibility". (for now it's written as BC, I guess that's why you thought it's not clear).

File: src/bootstrap/config.rs
Comment (ozkanonur) [reply src/bootstrap/config.rs:977]: Yeah, that's useless reorder. Will fix it.

File: src/bootstrap/config.rs
Comment (ozkanonur) [reply src/bootstrap/config.rs:805]: It's also needed in this PR. I can split https://github.com/rust-lang/rust/pull/107812/commits/a2f975a4f079ea1f1f0eeab5038f630d2595eb1e this commit out to another PR. But we will need it merged first before taking this one on master.

File: src/bootstrap/min_config.rs
Comment (Mark-Simulacrum) [reply src/bootstrap/min_config.rs:21]: Never seen it before, but I think in that case we need more here - what are the implications if we break compatibility? Shouldn't it be the same across both configs?

Comment (ozkanonur) [reply src/bootstrap/min_config.rs:21]: I see, will provide more detailed informations

File: src/bootstrap/config.rs
Comment (ozkanonur) [reply src/bootstrap/config.rs:751]: > We should always start with MinimalConfig::parse, and then take that as an argument for the big Config struct.

Right now `min_config` module is short and simple. If I apply the refactoring based on this suggestion, this will cause having a lot of `Config` field calculations in `min_config` module which will kill the seperation between them. Which will make the maintenance harder for both `Config` and `MinConfig`. Are you sure you want me to do this?

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs]: I was quite wrong about backwards compatibility. Just realised that.

https://github.com/rust-lang/rust/blob/d3039c7671c615a7f822279a19c0205b2a05bf7b/src/bootstrap/min_config.rs#L20-L24

Is this clear enough?

Comment (Mark-Simulacrum) [reply src/bootstrap/min_config.rs]: This is very much focusing on the literal usage of the struct, rather than why we have the separation and what reviewers/authors of changes should think through. If we just used Config, what would go wrong? I really want us to document that in the code, especially since it seems like we can't get a clear answer when asking the question now :) @jyn514 -- maybe you recall the details here?

My recollection is that we were concerned about the shim being out of date and not parsing config.toml as a result (i.e., a new field in config.toml would fail to parse with a slightly out of date shim). Is that the motivation here? It feels a little weak to me (we should be able to check that the shim people will be using is good enough for changes to config in CI, so it would just add some delay when adding and using new fields, I think?). But regardless, that should be clearly documented.

And, it seems like if we *are* worried about this kind of scenario, we should probably add a checkbox to https://github.com/rust-lang/rust/issues/94829 for validating it in CI.

But maybe I'm misremembering and we had some other scenario in mind.

File: src/bootstrap/config.rs
Comment (Mark-Simulacrum) [reply src/bootstrap/config.rs:751]: I don't understand. Why would Config field calculations need to move into MinConfig for this refactoring? It seems like only the calculations for the fields already in MinConfig would need to move; nothing else should. And those calculations already need to be in MinConfig anyway, as otherwise we can't just parse it?

Comment (ozkanonur) [reply src/bootstrap/config.rs:751]: So you want `MinConfig` field mapping to be moved in `config` module then. I am still unsure if this will be better. We intentionally did this seperation with 54b7ff25547f5893932ba0319234206a2d5fe2c4,  but I can do it if you want.

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs]: >  If we just used Config, what would go wrong?

I don't think things can go wrong, but it's unnecessary to parse whole config when you only need few of them for `bootstrap-shim` initialization.

I leave this to @jyn514 to give further details

Comment (Mark-Simulacrum) [reply src/bootstrap/min_config.rs]: Yes, I think we're missing something. We should not pay all this complexity for a few milliseconds less CPU time, if that.

Comment (jyn514) [reply src/bootstrap/min_config.rs]: Like you said, the performance wasn't really a consideration. There are two reasons I wanted to avoid parsing all of `Config`:
1. Unlike bootstrap itself, bootstrap-shim needs to be compatible across multiple versions of the Rust repo. I thought separating the fields it reads into another struct and adding a big "DO NOT CHANGE THIS LOGIC WITHOUT CHANGING THE SHELL SCRIPTS" comment would make it easier to maintain without accidentally regressing the shim.
2. When I wrote this originally, I wanted to make `bootstrap-shim` entirely separate from `bootstrap` itself except for the config.toml fields it reads, and maybe some things in `util.rs` and `build_helper`, with the goal of making it extremely fast to compile and download. Right now we're effectively downloading bootstrap twice, which is *fine*, but seems kind of wasteful. It ended up being a *lot* more work than I expected to separate them, but my hope was by at least separating the configs it will be easier to change the shim to not depend on all 90 of bootstrap's dependencies in the future.

If it's turning out to be a hassle, I don't mind using the same Config for both for now, but I think we should at least add the "DON'T TOUCH THESE FIELDS" comment and open a follow-up issue for making the shim smaller in the future.

Honestly if we're using the same config for both I'm not sure it helps to have a shim? at that point we may as well package bootstrap itself in each nightly, since they're effectively the same binary.

Comment (Mark-Simulacrum) [reply src/bootstrap/min_config.rs]: > Unlike bootstrap itself, bootstrap-shim needs to be compatible across multiple versions of the Rust repo. I thought separating the fields it reads into another struct and adding a big "DO NOT CHANGE THIS LOGIC WITHOUT CHANGING THE SHELL SCRIPTS" comment would make it easier to maintain without accidentally regressing the shim.

I think this is a good start to such a comment -- but I think we also should dig a little deeper in the comment. In particular, I think this "must be compatible" piece is a *distinct* requirement from the one we typically have -- usually we always are checking the current commit and searching back to find the latest change to some tool and then downloading that.

But, the plan for bootstrap-shim is, I guess, different : we will be using a *global* install of it? I forget the details now, but this seems like *exactly* the content I would want in a comment on the minimal config so that we can read that and understand what the compatibility story should be. I'm also not sure how the shell scripts play into this yet, but in any case good to reflect into the comment I'm looking for here.

I don't have a huge stake in what we ship or don't ship, but it seems to me that at least in theory if the shim is a separate binary we *should* see LLVM/linkers cut out any unused dependencies. On the other hand, maybe that doesn't give us that much - seems like the shim ends up at least pulling in decompression (xz + gzip?), tar, etc, so it probably won't be tiny.

Anyway to summarize it seems to me that probably the next step is for @jyn514 to maybe either point to a hackmd (I seem to recall one) or otherwise reflect some of these thoughts into the comment, so we can move forward based on that. I think there's also some other comments on this PR I've left that seem to still be warranted to resolve from my perspective, but it's possible that with a more detailed description of the purpose/rationale/strategy for the shim and minimal config I'll change my mind :)

Sounds like we're making definite progress though.

Comment (jyn514) [reply src/bootstrap/min_config.rs]: Here's the hackmd you're remembering: https://hackmd.io/@jynelson/S1Aa4cozo/edit

Comment (ozkanonur) [reply src/bootstrap/min_config.rs]: once the blocker PR is merged, I will re-update the doc-comments by the given updates from @jyn514 

> Here's the hackmd you're remembering: https://hackmd.io/@jynelson/S1Aa4cozo/edit

FYI the link is broken for me. Returns 404 Not Found.

Comment (jyn514) [reply src/bootstrap/min_config.rs]: fixed, thank you

File: src/bootstrap/check.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/check.rs:515]: This feels off to me - shouldn't checking bootstrap be enough? I would expect that we only need a single check target for each Cargo package, not for each binary target...

Comment (ozkanonur) [reply src/bootstrap/check.rs:515]: > This feels off to me - shouldn't checking bootstrap be enough?

No, sadly not enough. The same approach is applied for miri as well:
https://github.com/rust-lang/rust/blob/17675855094906ba90aca2f119be8fb7afc4d456/src/bootstrap/check.rs#L507-L508

Stdout after removing `tool_check_step!(BootstrapShim...`:

```sh
  ~/devspace/.other/rust-dist/general  $ ~/devspace/ozkanonur/rust/x check
Building bootstrap
   Compiling bootstrap v0.0.0 (/home/nimda/devspace/ozkanonur/rust/src/bootstrap)
error[E0412]: cannot find type `BootstrapShim` in module `check`
   --> builder.rs:660:24
    |
660 |                 check::BootstrapShim,
    |                        ^^^^^^^^^^^^^
    |
   ::: check.rs:418:9
    |
418 |         pub struct $name {
    |         ---------------- similarly named struct `Bootstrap` defined here
    |
help: a struct with a similar name exists
    |
660 |                 check::Bootstrap,
    |                        ~~~~~~~~~
help: consider importing this struct
    |
1   | use crate::dist::BootstrapShim;
    |
help: if you import `BootstrapShim`, refer to it directly
    |
660 -                 check::BootstrapShim,
660 +                 BootstrapShim,
    |

For more information about this error, try `rustc --explain E0412`.
error: could not compile `bootstrap` due to previous error
failed to run: /home/nimda/devspace/.other/rust-dist/general/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /home/nimda/devspace/ozkanonur/rust/src/bootstrap/Cargo.toml
Build completed unsuccessfully in 0:00:01
```

File: src/bootstrap/builder.rs
Comment (jyn514) [inline src/bootstrap/builder.rs:689]: looks like this is missing the new `BootstrapShim` step?

Comment (ozkanonur) [reply src/bootstrap/builder.rs:689]: Yeah was debugging it. We have problem at https://github.com/rust-lang/rust/blob/66f8dd12551d11eea0fbcf1bf95b2e18fb07a826/src/bootstrap/builder.rs#L519-L525

because it expects `src/bootstrap/bin/bootstrap-shim` which comes from:

```rs
tool_check_step!(BootstrapShim, "src/bootstrap/bin/bootstrap-shim", SourceType::InTree, false);
```

Comment (ozkanonur) [reply src/bootstrap/builder.rs:689]: Why not removing `tool_check_step!(BootstrapShim,..` as well ? Seems there is no problem at all with the `x check` command.

Comment (jyn514) [reply src/bootstrap/builder.rs:689]: As long as `x check bootstrap` also checks `bootstrap-shim`, that seems ok to me.

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/pull/107812#discussion_r1141060282 seems still unresolved here, but it may merit more synchronous discussion - I feel like we're not yet aligned in fact (even if there's no objections) on the exact trajectory here, and that makes me reluctant to merge PRs that create future implications (in particular I still don't think the comments align with a global install).

**Comment by ozkanonur** [comment]
> but it may merit more synchronous discussion

I agree, let's talk about this on the next team meeting.

File: src/bootstrap/min_config.rs
Comment (ozkanonur) [reply src/bootstrap/min_config.rs]: https://github.com/rust-lang/rust/blob/8ad89068e2558940b4184d46763ff9c6b325883b/src/bootstrap/min_config.rs#L20-L26

I didn't want to leave it as `to be filled`, because devs(who needs to make a change on config) might still be confused until we make the shim default. So I give the idea why it shouldn't be changed directly, and left a `FIXME` to update it once shim is default.

What do you think about the current change? @Mark-Simulacrum

**Pull request closed without merging**
